File: Miscellaneous/Python/155. Min Stack.py
class MinStack:

    def __init__(self):
        self.stack = []        

    def push(self, val):
        currMin = self.getMin()
        if currMin == None or val < currMin:
            currMin = val
        self.stack.append((val, currMin))

    def pop(self):
        self.stack.pop()
        

    def top(self):
        if len(self.stack) == 0:
            return None
        return self.stack[len(self.stack)-1][0]
        

    def getMin(self):
        if len(self.stack) == 0:
            return None
        return self.stack[len(self.stack)-1][1]
        


obj = MinStack()
obj.push(10)
obj.push(5)
#obj.pop()
obj.push(2)
obj.push(12)
param_3 = obj.top()
param_4 = obj.getMin()
print(param_3,  " ", param_4)

File: Miscellaneous/Golang/155. Min Stack.go
package main

import (
	"fmt"
)

type MinStack struct {
	Stack [][]int
}

func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(val int) {
	currMin := this.GetMin()
	if currMin == 0 || val < currMin {
		currMin = val
	}
	temp := make([]int, 2)
	temp[0], temp[1] = val, currMin
	this.Stack = append(this.Stack, temp)
}

func (this *MinStack) Pop() {
	this.Stack = this.Stack[:len(this.Stack)-1]
}

func (this *MinStack) Top() int {
	if len(this.Stack) == 0 {
		return 0
	}
	return this.Stack[len(this.Stack)-1][0]
}

func (this *MinStack) GetMin() int {
	if len(this.Stack) == 0 {
		return 0
	}
	return this.Stack[len(this.Stack)-1][1]
}

func main() {
	obj := Constructor()
	obj.Push(10)
	obj.Push(5)
	obj.Pop()
	obj.Push(2)
	obj.Push(12)
	param_3 := obj.Top()
	param_4 := obj.GetMin()
	fmt.Println(param_3, " ", param_4)
}
